packages/flutter_tools/lib/src/features.dart

```diff
@@ -0,0 +1,245 @@
+      }
+    }
+    if (feature.environmentOverride != null) {
+      if (platform.environment[feature.environmentOverride]?.toLowerCase() == 'true') {
+        isEnabled = true;
+      }
+    }
+    return isEnabled;
+  }
+}
+
+/// A [Feature] is a process for conditionally enabling tool features.
+///
+/// All settings are optional, and if not provided will generally default to
+/// a "safe" value, such as being off.
+///
+/// The top level feature settings can be provided to apply to all channels.
+/// Otherwise, more specific settings take precidence over higher level
+/// settings.
+class Feature {
+  /// Creates a [Feature].
+  const Feature({
+    @required this.name,
+    this.environmentOverride,
+    this.configSetting,
+    this.master = const FeatureSetting(),
+    this.dev = const FeatureSetting(),
+    this.beta = const FeatureSetting(),
+    this.stable = const FeatureSetting(),
+  });
+
+  /// The user visible name for this feature.
+  final String name;
+
+  /// The settings for the master branch and other unknown channels.
+  final FeatureSetting master;
+
+  /// The settings for the dev branch.
+  final FeatureSetting dev;
+
+  /// The settings for the beta branch.
+  final FeatureSetting beta;
+
+  /// The settings for the stable branch.
+  final FeatureSetting stable;
+
+  /// The name of an environment variable that can override the setting.
+  ///
+  /// The environment variable only needs to be "set", that is contain a
+  /// non empty string. This is only intended for usage by CI and not
+  /// as an advertised method to enable a feature.
+  ///
+  /// If not provided, defaults to `null` meaning there is no override.
+  final String environmentOverride;
+
+  /// The name of a setting that can be used to enable this feature.
+  ///
+  /// If not provided, defaults to `null` meaning there is no config setting.
+  final String configSetting;
+
+  /// A help message for the `flutter config` command, or null if unsupported.
+  String generateHelpMessage() {
+    if (configSetting == null) {
+      return null;
+    }
+    final StringBuffer buffer = StringBuffer('Enable or disable $name on ');
+    final List<String> channels = <String>[
+      if (master.available) 'master',
+      if (dev.available) 'dev',
+      if (beta.available) 'beta',
+      if (stable.available) 'stable',
+    ];
+    if (channels.length == 1) {
+      buffer.write('the ${channels.single} channel.');
+    } else {
+      buffer.write('${channels.join(', ')} channels.');
+    }
+    return buffer.toString();
+  }
+
+  /// Retrieve the correct setting for the provided `channel`.
+  FeatureSetting getSettingForChannel(String channel) {
+    switch (channel) {
+      case 'stable':
+        return stable;
+      case 'beta':
+        return beta;
+      case 'dev':
+        return dev;
+      case 'master':
+      default:
+        return master;
+    }
+  }
+}
+
+/// A description of the conditions to enable a feature.
+class FeatureSetting {
```

Comment:
Are these things ever used independently? Otherwise, just name them FeatureChannelSetting for clarity?

Reply:
Done